### PR TITLE
[remove nixpkgs] prototype for linux x86

### PR DESCRIPTION
**Pull request by savil**
## Summary

This PR:

- [x] generate `flake_alt.nix.tmpl` for user's system only 
- [ ] incorporate `toPath` to avoid `--impure`
- [ ] try moving `shellgen.Package` to `nix.Input`

Follow up PRs:
- [ ] store fetchClosureArgs in the lockfile file 
- [ ] `flake_alt.nix.tmpl` should work for all platforms
- [ ] support php packages
- [ ] support local flake packages
- [ ] support remote flake packages
- [ ] find nixpkgs commit hash installed with nix and use that in `flake_alt.nix.tmpl`

## How was it tested?

In devbox-cloud's browser, I was able to:
- do some hackery to disable `devbox global` and the shellenv environment to get a clean bash shell.
- `export DEVBOX_FEATURE_REMOVE_NIXPKGS=1`.
- cd `examples/development/go/hello-world`.
- `devbox shell` and in it do `go run .`

Took > 10 seconds when starting `devbox shell` when fetching `fetchClosureArgs` from search api during shell invocation. Hence the TODO above to store `fetchClosureArgs` during `devbox.PrintEnv`.

### CODE REVIEW

**Comment by savil** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1182** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/1182" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jetpack-io/devbox/1182?utm_source=stack-comment).

File: internal/planner/plansdk/plansdk.go [hunk 1]
```diff
@@ -7,16 +7,40 @@ import (
...
 )
 
 // FlakePlan contains the data to populate the top level flake.nix file
 // that builds the devbox environment
 type FlakePlan struct {
 	NixpkgsInfo *NixpkgsInfo
 	FlakeInputs []*FlakeInput
+
+	// Packages are used by "RemoveNixpkgs" feature
+	Packages []*Package
```
Comment (mikeland73) [inline internal/planner/plansdk/plansdk.go:27]: I would really prefer you reuse `FlakeInputs`. This is basically the same. You can rename it to packages if you want (but that's a riskier change because you need to make sure you don't break templates.

**Review by mikeland73 — COMMENTED** [review]
Nice draft!

I think we should:

* Merge new package struct into `nix.Input`
* Have lock.Resolve fetch store paths and save them to lockfile for perf. If that bloats the PR you could always just have it fetch the store path with feature flag and do the lockfile manipulation in follow up.
* Ensure we are using correct search endpoint. What we need is an endpoint that given and ID will return store paths. (Or we can reuse `SearchVersion` will already returns a semi-deterministic exact version. We can add new fields to this cc: @gcurtis ).

Two notes to expand on requests:

* `nix.Input` is supposed to be renamed to Packages in town package (possibly `devpkg.Package`) I chose the name `Input` previously to avoid colliding too much with string packages we were using previously. Now that almost everything is migrated to Input, we could make that change. Creating a `Package` struct would be counter to this.
* Moving the resolution to the lockfile is pretty important. It's the one centralized place were we do search calls to fetch package details. Since it has built in caching it allows for better performance for all users (shell, run, etc). It also has `update` built in.

File: internal/impl/generate.go [hunk 2]
```diff
@@ -146,7 +146,12 @@ var templateFuncs = template.FuncMap{
 
 func makeFlakeFile(outPath string, plan *plansdk.FlakePlan) error {
 	flakeDir := filepath.Join(outPath, "flake")
-	err := writeFromTemplate(flakeDir, plan, "flake.nix")
+
+	templateName := "flake.nix"
+	if featureflag.RemoveNixpkgs.Enabled() {
+		templateName = "flake_alt.nix"
```
[the remark this answers is not in the document]
Comment (savil) [reply internal/impl/generate.go:152]: yes, renaming to a descriptive name had been the plan

**Review by savil — COMMENTED** [review]
Thanks for the comments! Bear in mind that this was a proof-of-concept draft to validate that `builtins.fetchClosure` could work. It was not ready for review.

> I think we should:
Merge new package struct into nix.Input

Will try, but to be honest may also keep separate if its too hard. Will bias towards moving forward, and we can clean up code later. The current `nix.Input` code is fairly hard to follow for me.

> Have lock.Resolve fetch store paths and save them to lockfile for perf. If that bloats the PR you could always just have it fetch the store path with feature flag and do the lockfile manipulation in follow up.

Yes, we do need the lockfile to store the `fromPath` and `toPath` info. There will be many, many follow up PRs fear not. I don't believe in massive PRs ;-)

> Ensure we are using correct search endpoint. What we need is an endpoint that given and ID will return store paths. (Or we can reuse SearchVersion will already returns a semi-deterministic exact version. We can add new fields to this cc: @gcurtis ).

The search endpoint was suggested by @gcurtis but I also share concerns about ensuring we get the `fromPath` for the exact version of the package we care about. Will touch base with @gcurtis IRL.

> nix.Input is supposed to be renamed to Packages in town package (possibly devpkg.Package) I chose the name Input previously to avoid colliding too much with string packages we were using previously. Now that almost everything is migrated to Input, we could make that change. Creating a Package struct would be counter to this.

Gotcha. That's good context. I'm okay with keeping `input` as a name. I chose "Package" because its unused currently :p


> Moving the resolution to the lockfile is pretty important. It's the one centralized place were we do search calls to fetch package details. Since it has built in caching it allows for better performance for all users (shell, run, etc). It also has update built in.

yes, agreed on lockfile, but disagreed on the reason. If it were just perf, then we can save the `fetchClosureArgs` info in a cache in `.devbox` directory. The reason to add to lockfile is so that the search-service is not on the critical path for users to build the "devbox shell environment" that is needed for devbox run/shell/shellenv.

File: internal/searcher/model.go [hunk 3]
```diff
@@ -29,3 +29,28 @@ type SearchResult struct {
 	Results     []Result `json:"results"`
 	Suggestions []Result `json:"suggestions"`
 }
+
+// Package api:
+// https://search.devbox.sh/pkg/<name>
```
[the remark this answers is not in the document]
Comment (gcurtis) [reply internal/searcher/model.go:34]: Yeah, I think it would be a good idea to allow lookup by ID. The information needed to unambiguously identify a package is name, version, and system.

The `/pkg/<name>` endpoint will get you the correct information, but the client does have to iterate through the results to find the right name + version + system combination itself, which isn't ideal.

[the remark this answers is not in the document]
Comment (savil) [reply internal/searcher/model.go:34]: Discussed with @gcurtis 

1. He'll add the required information to the old search api endpoint. That way, we'll get the extra info this `/pkg` api provides during the same `devbox add` call.
2. For now, I can use the new endpoint to continue working on this.

**Comment by savil** [comment]
PR stack starting at #1208 implements the feature.
